src/cuda/graph_ops.cu

```diff
@@ -1,77 +1,133 @@
 #include "graph_ops.h"
 
+#include "atomic.h"
 #include "cuda_common.h"
 #include "utils.h"
 
 namespace gs {
 namespace impl {
+template <typename IdType, typename DType>
+__global__ void GroupSum(IdType* indptr, IdType* e_ids, DType* data,
```

Comment:
Would be better to rename it to `SegmentSum`. I believe this function is the same as segment sum in thrust/cub. The only difference is you may have id mapping here. To make thing simple, I think you can apply a id permutation if necessary and then use the segment sum in thrust/cub.